networkit/cpp/generators/RmatGenerator.cpp

```diff
@@ -12,17 +15,155 @@
+        return result;
+    }
+};
+
+void generateEntries(std::vector<Entry> &entryList, count size, double a, double b, double c,
+                     double d) {
+    entryList.reserve(size);
+    PriorityQueue priorityQueue(entryList);
+    priorityQueue.push(Entry{0, 0, 1, a});
+    priorityQueue.push(Entry{0, 1, 1, b});
+    priorityQueue.push(Entry{1, 0, 1, c});
+    priorityQueue.push(Entry{1, 1, 1, d});
+    while (entryList.size() <= size - 3) {
+        // Take the entry with the highest probability and split it up.
+        // This increases the average number of bits compared to a uniform distribution.
+        Entry old = priorityQueue.pop();
+        priorityQueue.push(
+            Entry{old.i << 1 | 0, old.j << 1 | 0, old.numberOfBits + 1, a * old.probability});
+        priorityQueue.push(
+            Entry{old.i << 1 | 0, old.j << 1 | 1, old.numberOfBits + 1, b * old.probability});
+        priorityQueue.push(
+            Entry{old.i << 1 | 1, old.j << 1 | 0, old.numberOfBits + 1, c * old.probability});
+        priorityQueue.push(
+            Entry{old.i << 1 | 1, old.j << 1 | 1, old.numberOfBits + 1, d * old.probability});
+    }
+}
+
 RmatGenerator::RmatGenerator(count scale, count edgeFactor, double a, double b, double c, double d,
-                             bool weighted, count reduceNodes)
-    : scale(scale), edgeFactor(edgeFactor), a(a), b(b), c(c), weighted(weighted),
-      reduceNodes(reduceNodes) {
+                             bool weighted, count reduceNodes, bool discardSelfLoops)
+    : scale(scale), edgeFactor(edgeFactor), weighted(weighted), reduceNodes(reduceNodes),
+      discardSelfLoops(discardSelfLoops) {
     if (scale > 63)
         throw std::runtime_error("Cannot generate more than 2^63 nodes");
     double sum = a + b + c + d;
     INFO("sum of probabilities: ", sum);
     if (!Aux::NumericTools::equal(sum, 1.0, 0.0001))
         throw std::runtime_error("Probabilities in Rmat have to sum to 1.");
     defaultEdgeWeight = 1.0;
+
+    count size = 1 << std::min((count)12, (count)std::round(scale * 5.0 / 8.0) + 1);
+
+    assert((size & (size - 1)) == 0); // The size must be a power of two.
+    std::vector<Entry> entryList;
+    generateEntries(entryList, size, a, b, c, d);
+    while (entryList.size() < size) { // Fill the remaining entries with 0-probability entries.
+        entryList.push_back(Entry{0, 0, 0, 0});
+    }
+
+    // Construct the alias table:
+    mask = size - 1;
+    bits.resize(size);
+    numberOfBits.resize(size);
+    coinFlipProbability.resize(size);
+    coinFlipReplacement.resize(size);
+    for (count i = 0; i < size; i++) {
+        bits[i] = std::make_pair(entryList[i].i, entryList[i].j);
+        numberOfBits[i] = entryList[i].numberOfBits;
+        coinFlipProbability[i] = 0;
+        coinFlipReplacement[i] = 0;
+    }
+    double baseProbability = 1.0 / (double)size;
+    count lastOverfullIndex = 0;
+    count lastUnderfullIndex = 0;
+    while (true) {
```

Comment:
Avoid using `while(true)` for both readability and maintainability. Use `do {} while(...)` instead.